File: Leetcode/Edit Distance/Edit Distance.java
class Solution {
    public int minDistance(String word1, String word2) {
        //Finding size of both strings
        int m = word1.length();
        int n = word2.length();
        
        //Initialsing dp
        int[][] dp = new int[m + 1][n + 1];
        
        //If either of strings are empty, then each letter will have to be deleted/added. This is the base case
        for(int i = 0; i <= m; i++){
            dp[i][0] = i;
        }
        for(int i = 1; i <= n; i++){
            dp[0][i] = i;
        }
        
        //Looping through the strings
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                //If the characters are equal, then the operations will be same as previous operations
                if(word1.charAt(i) == word2.charAt(j))
                    dp[i + 1][j + 1] = dp[i][j];
                //If they are not equal, then we will select the cheapest operation, i.e. insert, delete or replace and that will be our ans
                else {
                    int a = dp[i][j];
                    int b = dp[i][j + 1];
                    int c = dp[i + 1][j];
                    dp[i + 1][j + 1] = a < b ? (a < c ? a : c) : (b < c ? b : c);
                    dp[i + 1][j + 1]++;
                }
            }
        }
        //Return the ans
        return dp[m][n];
    }
}

File: Leetcode/Edit Distance/Edit Distance.py
"""
Given two strings word1 and word2, return the minimum number of operations required to convert word1 to word2.

You have the following three operations permitted on a word:

Insert a character
Delete a character
Replace a character

"""

class Solution:
    def minDistance(self, word1, word2):
        s1, s2 = len(word1), len(word2)  #Calculate length of word1 & word2 for forming dp grid
        dp = [[0 for _ in range(s2+1)] for _ in range(s1+1)] # Form a grid corresponding to length of strings
        # Here n & m are the length of the strings
        def helper(n, m):
            # If there is no character in word1 then to make it word2, len(word2) characters needs to be added.
            if n==0:
                return m
             # If there is no character in word2 then to make word1 equal to word2, len(word1) characters needs to be deleted.
            if m==0:
                return n
            if dp[n][m]:
                return dp[n][m]
            # If the last charcter are equal then run resursive for n-1,m-1 value
            if word1[n-1] == word2[m-1]:
                dp[n][m] = helper(n-1, m-1)
                return dp[n][m]
            # [helper(n-1, m): Delete a charcater], [helper(n-1, m-1): Subtitution], [helper(n, m-1): Addition]
            ans = 1 + min(helper(n, m-1), helper(n-1, m-1), helper(n-1, m))
            dp[n][m] = ans
            return ans
        
        return helper(s1, s2)

    